### DISCUSSION TITLE
[Suggestions] Allow to use the mouse right click anywhere on the screen when a window is open, and the "ESC" key on the keyboard, to close a window with an "OKAY" button and with a "CANCEL" button

### DISCUSSION THREAD

**Discussion by LeHerosInconnu** [ihhub/fheroes2#2607] [Ideas]
Allow to use the mouse right click anywhere on the screen when a window is open, and the "ESC" key on the keyboard, to close a window with an "OKAY" button and with a "CANCEL" button.

In the original game, this operating mode with the "ESC" key is already available to close windows with an "OKAY" button (when this button is the only one) and windows with a "CANCEL" button (as if the user pressed the "CANCEL" button).

**Comment by Branikolog** [comment]
If it's original operation of Esc buton, why haven't you created an issue already, @LeHerosInconnu? :)
Also, you didn't specified a mouse operation here. In which cases we want to allow mouse clicking?

**Reply by LeHerosInconnu** [reply]
@Branikolog.

> If it's original operation of Esc buton, why haven't you created an issue already, @LeHerosInconnu? :)

Because I thought it would be a good idea to do both at the same time and there are already a lot of issues waiting to be resolved. :)

> Also, you didn't specified a mouse operation here. In which cases we want to allow mouse clicking?

If there are possible actions with the right click of the mouse in the screens/windows, they are limited to the areas that allow these actions and they must therefore have priority over the "general" right click proposed here that allows to close the window.

**Reply by Branikolog** [reply]
Ah, I've a little misunderstanding here... 
You are proposing to close all dialog windows by right click on an outer area, right?
I think this improvement is questionable... As player can misclick and not able to read the message on the screen... I prefer leaving only original behaviour, where we can close any window by "distant" and "not-pressed-accidentally" buttnon "Esc". :)

**Reply by LeHerosInconnu** [reply]
@Branikolog.

> Ah, I've a little misunderstanding here...
> You are proposing to close all dialog windows by right click on an outer area, right?

Anywhere on the screen, except on active areas for a mouse button.

> I think this improvement is questionable... As player can misclick and not able to read the message on the screen...

This can also be an option for people who often make misclicks :)
And clicking outside the window to avoid clicking errors can also be a practice to adopt.

> I prefer leaving only original behaviour, where we can close any window by "distant" and "not-pressed-accidentally" buttnon "Esc". :)

The goal is to propose the same action with the mouse.
The "ESC" key is not easily reached with the mouse. :D

Personally I prefer to use the mouse rather than the keyboard to perform actions.
It's more natural and convenient than moving the hand on the specific key of the keyboard.
And there are a lot of people in this case.

That's why any program that is worthy should offer to perform actions in several ways, here with the mouse and with the keyboard.
It is up to the user to choose the way that suits him best.

**Reply by Branikolog** [reply]
Humm. Actually cannot remember any similar mechanics anywhere in my memory... Regarding any software or games...
Probably, for some confirmation dialogs it could be useful... I mean those, you could reproduce by clicking any UI buttons or places...
But for some text boxes with only "OK" button we can miss acidentaly some useful information...
I think we should at least allow "Esc" button at such cases first. And then debate about this specific mouse click operation...

**Reply by LeHerosInconnu** [reply]
@Branikolog.

> But for some text boxes with only "OK" button we can miss acidentaly some useful information...

So, you think it's necessary to add it as an option?

**Reply by Branikolog** [reply]
Actually no.)
We could polish the game and invent new interaction methods for eternity. I think we should add only the most useful and "musthave" features, as adding also requires time, which we can use for improving game content.

**Reply by LeHerosInconnu** [reply]
@Branikolog.

> Actually no.)
> We could polish the game and invent new interaction methods for eternity. I think we should add only the most useful and "musthave" features, as adding also requires time, which we can use for improving game content.

I'm not a programmer, but I doubt it would take a lot of time to add a feature like this. :)

**Reply by Branikolog** [reply]
So do I.) 
I hope we will have much more people contributing to the project some time... 
For now, our team has barely free time to analyse bugs, we both create in "issues". :)
Let's leave this mouse improvement for discussion... 
Also, I 100% would like to use "Esc" button in a way, you've proposed. I think, we should create an issue later for it...

**Reply by LeHerosInconnu** [reply]
@Branikolog.

> So do I.)
> I hope we will have much more people contributing to the project some time...
> For now, our team has barely free time to analyse bugs, we both create in "issues". :)
> Let's leave this mouse improvement for discussion...
> Also, I 100% would like to use "Esc" button in a way, you've proposed. I think, we should create an issue later for it...

It's true that the issue counter has a hard time going below 500, although I've slowed down my bug reports considerably.
And I don't really see the point of raising it to 600, 800 or 1000 issues.
That would discourage people more than anything else I think.

**Reply by Branikolog** [reply]
I believed, that slow down was due the fact, we've found the most of them... 
Don't tell me, that you're holding a few hundreds of them... T_T

**Reply by LeHerosInconnu** [reply]
@Branikolog.

> I believed, that slow down was due the fact, we've found the most of them...
> Don't tell me, that you're holding a few hundreds of them... T_T

I'm not going to say it then. :)
